src/wallet.ts

```diff
@@ -124,3 +125,52 @@ export class Wallet extends AdapterL2(AdapterL1(ethers.Wallet)) {
         return await this.provider.broadcastTransaction(await this.signTransaction(populatedTx));
     }
 }
+
+interface SigningFunction {
+    (this: AbstractWallet, transaction: TransactionRequest): Promise<string>;
+}
+
+export class AbstractWallet extends Wallet {
+    readonly accountAddress: string;
+    paymaster: Paymaster;
+    customSigningFunction: SigningFunction;
+
+    constructor(
+        accountAddress: string,
+        privateKey: string,
+        providerL2?: Provider,
+        providerL1?: ethers.Provider,
+        paymaster: Paymaster | null = null,
+        customSigningFunction: SigningFunction = signingFunction,
+    ) {
+        super(privateKey, providerL2);
+        this.accountAddress = accountAddress;
+        this.paymaster = paymaster;
+        this.providerL1 = providerL1;
+        this.customSigningFunction = customSigningFunction.bind(this);
+    }
+
+    override transfer(transaction: { to: string; amount: ethers.BigNumberish; token?: string; overrides?: ethers.Overrides; }): Promise<TransactionResponse> {
+        transaction.overrides = { type: EIP712_TX_TYPE };
+        return super.transfer(transaction);
+    }
+
+    override getAddress(): Promise<string> {
+        return Promise.resolve(this.accountAddress);
+    }
+
+    override async signTransaction(transaction: TransactionRequest): Promise<string> {
+        return this.customSigningFunction(transaction);
+    }
+}
+
+export async function signingFunction(this: AbstractWallet, transaction: TransactionRequest): Promise<string> {
+    transaction.customData = this.paymaster ? {
+        gasPerPubdata: DEFAULT_GAS_PER_PUBDATA_LIMIT,
+        paymasterParams: this.paymaster.getPaymasterParams(),
+    } : {};
+    transaction.customData = this._fillCustomData(transaction.customData);
+    transaction.customData.customSignature = await this.eip712.sign(transaction);
```

Comment:
how does it work? What key is used to sign it?

Reply:
This will use the `privateKey` parameter passed to the constructor here. Ultimately, a `SigningKey` function is added at `ethers.BaseWallet`, the class that `ethers.Wallet` extends (and this class is extended by `zks.Wallet`), using this passed in private key that is used for signatures. 

I find it also tricky to know what exactly is going on with the class as it extends another class about 5 times.. But this approach adds the least lines of code too.